### PR TITLE
Change endpoint from str to Url

### LINKED ISSUE #3: Change endpoint from str to URL?
at the moment something like:
`bitcoin:TB1Q6D3A2W975YNY0ASUVD9A67NER4NKS58FF0Q8G4?amount=1&pj=http://a`
is parsed as a valid bip78 URI, did you consider having field `endpoint` in `Uri` struct to be a parsed URL instead of str (even at the cost of adding a dep)?

### PR DESCRIPTION
**Pull request by DanGould**
Close #3

I question whether I did right by removing deserialize_borrow.

As an aside, bip21 looks like it will be accepted into rust-bitcoin now. percent-encoding seems to be part of that work rather than this.

### CODE REVIEW

**Review by Kixunil — CHANGES_REQUESTED** [review]
Thanks! Yeah, removing the lifetime and `deserialize_borrowed` was correct since `Url` can't reuse the buffer. I was thinking that a new string parameter might be added in the future but it seems unlikely and not worth having more lifetimes.

Panic must be definitely fixed, it's a DoS vector at worst.

File: bip78/src/uri.rs [hunk 1]
```diff
@@ -121,30 +123,13 @@ impl<'a> bip21::de::DeserializationState<'a> for DeserializationState<'a> {
...
-            _ => Ok(bip21::de::ParamKind::Unknown),
-        }
-    }
-
-    fn deserialize_borrowed(&mut self, key: &'a str, value: bip21::Param<'a>) -> std::result::Result<bip21::de::ParamKind, <Self::Value as bip21::DeserializationError>::Error> {
-        match key {
-            "pj" if self.pj.is_none() => {
-                self.pj = Some(value.try_into().map_err(InternalPjParseError::NotUtf8)?);
+                self.pj = Some(
+                    Url::parse(&*Cow::try_from(value).map_err(InternalPjParseError::NotUtf8)?)
```
[the remark this answers is not in the document]
Comment (evanlinjin) [reply bip78/src/uri.rs:130]: I also want to point out you don't technically need the deref `*` either. From my understanding, it basically forces ownership?

**Review by evanlinjin — CHANGES_REQUESTED** [review]
Looks pretty good to me (based on my limited understanding of bip-78). Also, it would be cool to run `cargo fmt` before committing :)

**Comment by DanGould** [comment]
Panic, lifetime, style issues fixed. Ensure only  `http://....onion`s are allowed.

> Use it to manipulate query parameters (instead of hard-coding it). – @evanlinjin

I was looking for this yesterday, how'd you find it 😎 thanks, done.

> I also want to point out you don't technically need the deref * either. From my understanding, it basically forces ownership? – @evanlinjin 

We do need the deref  `*` to `match` on the _values_ `"0"` and `"1"`. Sidenote: It would have been nicer if the spec didn't use `pjos=0` and let us check for the mere presence of`dpjos` instead of parsing the argument. they make it worse by suggesting to use `disablepaymentsubsitution=true` so we've got digits and strings to parse. Inconsistency is a bother. I don't know of any implementation in the wild actually taking advantage of either, so perhaps the spec can be changed / version upgraded to check for boolean presence instead of parsing.

I have applied select cargo format changes to those within the scope of this review

File: bip78/src/uri.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply bip78/src/uri.rs]: It's just deref coercions.

**Comment by Kixunil** [comment]
Took only a quick look, have to run, should be back in a few hours.

**Review by Kixunil — APPROVED** [review]
Loogs good! @evanlinjin would you like to review again?

**Review by evanlinjin — APPROVED** [review]
Everything looks good!

Nit: It seems there are unused imports? I personally use `cargo clippy`, but `cargo fix` can work as well.

**Comment by DanGould** [comment]
Thank you for your reviews ⚡️

If there were a merge button for me to press I would press it, but I have no such ability

**Comment by Kixunil** [comment]
Oh, forgot to click, had too much stuff yesterday. Thanks again!

**Pull request merged**